docs/python-grader/index.md

```diff
@@ -0,0 +1,165 @@
+  <pl-file-editor
+    file_name="user_code.py"
+    ace_mode="ace/mode/python"
+    source-file-name="tests/initial_code.py"></pl-file-editor>
+</pl-question-panel>
+
+<pl-submission-panel>
+  <pl-external-grader-results />
+</pl-submission-panel>
+```
+
+### `tests/setup_code.py`
+
+This file is executed before any reference or student code is run.  Any variables defined in `names_for_user` can be accessed from here in student code, while the reference answer may freely access variables without restriction.
+
+### `tests/test.py`
+
+The test cases for each coding problem are defined as methods of a `test` class contained in the aptly named `test.py` file.  The class will have one of the following signatures, depending if you require plots from the student:
+
+```python
+## No plot grading
+class Test(PrairieLearnTestCase):
+
+## Plot grading enabled
+class Test(PrairieLearnTestCaseWithPlot):
+```
+
+Each test case is a separate method in the class, the names of these functions must be prefixed with `test_`, but any name that follows is arbitrary.  Test cases are run in sorted order by the `unittest` library, so one convention that can be used is to give tests numeric names to ensure their running order.
+
+Adding a name and point value to the test case is done by means of python decorators:
+`@points(val)` and `@name("name of the test case")`.  These will control the name of the case and the points awarded in the "Test Results" dropdown menu that are shown when a student submits a solution.  An example of a definition:
+
+```python
+@points(10)
+@name("Cool test case")
+def test_0(self):
+```
+
+Inside the test case implementation, the student answer variables and reference answer variables can be accessed as children of the tuples `self.st` and `self.ref`, respectively.  There are various helper functions to check correctness of different types of variables, these are defined in `code_feedback.py`.  These are taken from the RELATE grader, so this may be familiar to those with prior experience with RELATE.
+
+At the end of the test case, set the correctness of the answer using `feedback.set_points()`.  This function takes a floating point number between 0 and 1 (inclusive), with 0 being completely *in*correct and 1 being completely correct.  By default, if no points are given the test case will be marked incorrect.
+
+The overall structure of a test case should look something like:
+
+```python
+@points(10)
+@name("name of the test case")
+def test_0(self):
+   if feedback.check_scalar('name of the variable', self.ref.variable_name, self.st.variable_names):
+       Feedback.set_points(1)
+   else:
+       Feedback.set_points(0)
+```
+
+Note that `Feedback.set_points()` is used to set the _percentage_ correctness of the test case.  For example, if a test case is worth 10 points and `Feedback.set_points(0.5)` is run, the student will be awarded 5 points.
+
```

Comment:
Consider adding here a section on preventing different functions from being accessed or used with:

```py
la.inv = not_allowed
```

Out of pure curiosity, how deep is this function signature check performed? Could I get around it with:

```py
from numpy.linalg import inv as toad 
```

Reply:
![image](https://user-images.githubusercontent.com/1790491/80391606-93c50280-8873-11ea-8730-3c60a8d2a5dd.png)

Nope, I think that any modules that get imported in the setup code will get cached, and if the student's code tries to "import" it again it will use the hooked copy.